From e953e0e17a8a9d8da21164c67546ebf391839509 Mon Sep 17 00:00:00 2001
From: Xinhui Yang <cyan@cyano.uk>
Date: Tue, 31 Dec 2024 17:29:01 +0800
Subject: [PATCH 5/5] chore: use /boot/u-boot as the boot filesystem

Using /boot with FAT32 causes problems for dpkg. We now require the
u-boot filesystem to be mounted at /boot/u-boot.
---
 etc/kernel/postinst.d/zz-u-boot-menu |  2 +-
 read-config                          | 28 ++++++++++++++++++++++++++--
 u-boot-update                        | 10 +++++++++-
 3 files changed, 36 insertions(+), 4 deletions(-)

diff --git a/etc/kernel/postinst.d/zz-u-boot-menu b/etc/kernel/postinst.d/zz-u-boot-menu
index 7ef7b0e..db5348d 100755
--- a/etc/kernel/postinst.d/zz-u-boot-menu
+++ b/etc/kernel/postinst.d/zz-u-boot-menu
@@ -9,7 +9,7 @@ srcdir="/usr/lib/aosc-os-${ARCH}-boot/dtbs-kernel-${version}"
 destdir="${_BOOT_PATH}${_BOOT_DIRECTORY}${U_BOOT_FDT_DIR}${version}"
 
 # Sync DTBs if /boot is on a different partition than /
-if has_separate_boot && [ "${U_BOOT_SYNC_DTBS}" = "true" ] && command -v rsync >/dev/null 2>&1
+if u_boot_mounted && [ "${U_BOOT_SYNC_DTBS}" = "true" ] && command -v rsync >/dev/null 2>&1
 then
 	if [ -d "${srcdir}" ]
 	then
diff --git a/read-config b/read-config
index 0bd0a11..8a8b1d3 100644
--- a/read-config
+++ b/read-config
@@ -6,6 +6,10 @@ has_separate_boot() {
     [ "$(stat --printf %d /)" != "$(stat --printf %d /boot)" ]
 }
 
+u_boot_mounted() {
+	grep -q '[[:space:]]\+/boot/u-boot[[:space:]]\+' /proc/mounts
+}
+
 # Reading the default file
 if [ -e /etc/default/u-boot ]
 then
@@ -44,9 +48,29 @@ fi
 # Find boot directory as seen in u-boot, and path prefix while in linux
 if has_separate_boot
 then
-	# / and /boot are on different filesystems
+	echo "E: AOSC OS does not support /boot as the U-Boot filesystem."
+	echo "   Please mount the filesystem to /boot/u-boot, and try again."
+	echo "W: Remember to modify your /etc/fstab."
+	exit 1
+# In case of fstab is not generated yet.
+elif u_boot_mounted
+then
+	UBOOT_MOUNTED=1
+	_BOOT_DIRECTORY=""
+	_U_BOOT_DIRECTORY="/boot/u-boot/extlinux"
+	_BOOT_PATH="/boot/u-boot/"
+	if [ "${U_BOOT_SYNC_DTBS}" = "true" ]
+	then
+		_FDT_DIR="/dtb-"
+	fi
+elif grep -q /boot/u-boot /etc/fstab
+then
+	echo "W: /boot/u-boot is not mounted. Mounting ..."
+	mount /boot/u-boot || { echo "E: Failed to mount /boot/u-boot. Exiting." ; exit 1 ;}
+	UBOOT_MOUNTED=1
 	_BOOT_DIRECTORY=""
-	_BOOT_PATH="/boot"
+	_U_BOOT_DIRECTORY="/boot/u-boot/extlinux"
+	_BOOT_PATH="/boot/u-boot"
 	if [ "${U_BOOT_SYNC_DTBS}" = "true" ]
 	then
 		_FDT_DIR="/dtb-"
diff --git a/u-boot-update b/u-boot-update
index b7778c4..22cc888 100755
--- a/u-boot-update
+++ b/u-boot-update
@@ -120,19 +120,27 @@ do
 	# Strip kernel prefix to derive version.
 	_VERSION=${_KERNEL#*-}
 	echo "P: Writing config for ${_KERNEL}..."
-
+	(($UBOOT_MOUNTED)) && install /boot/${_KERNEL} \
+		/boot/u-boot/${_KERNEL}
 	_NUMBER="${_NUMBER:-0}"
 	_ENTRY="${_ENTRY:-1}"
 
 	if [ -e "/boot/${U_BOOT_INITRD}-${_VERSION}" ]
 	then
 		_INITRD="initrd ${_BOOT_DIRECTORY}/${U_BOOT_INITRD}-${_VERSION}"
+		(($UBOOT_MOUNTED)) && install "/boot/${U_BOOT_INITRD}-${_VERSION}" \
+			/boot/u-boot/"${U_BOOT_INITRD}-${_VERSION}"
+
 	elif [ -e "/boot/${U_BOOT_INITRD}-${_VERSION}.img" ]
 	then
 		_INITRD="initrd ${_BOOT_DIRECTORY}/${U_BOOT_INITRD}-${_VERSION}.img"
+		(($UBOOT_MOUNTED)) && install "/boot/${U_BOOT_INITRD}-${_VERSION}.img" \
+			/boot/u-boot/"${U_BOOT_INITRD}-${_VERSION}.img"
 	elif [ -e "/boot/${U_BOOT_INITRD}" ]
 	then
 		_INITRD="initrd ${_BOOT_DIRECTORY}/${U_BOOT_INITRD}"
+		(($UBOOT_MOUNTED)) && install "/boot/${U_BOOT_INITRD}" \
+			/boot/u-boot/"${U_BOOT_INITRD}"
 	else
 		_INITRD=""
 	fi
-- 
2.47.1

